### DISCUSSION TITLE
How do I add an external stylesheet from a CDN when using app-router?

### DISCUSSION THREAD

**Discussion by Svish** [vercel/next.js#49014] [Help] [answered]
### Summary

I need to add the following to one of my pages, but can't figure out how from the current app-router documentation.

```html
<link
  rel="stylesheet"
  href="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.css"
  integrity="sha384-Xi8rHCmBmhbuyyhbI88391ZKP2dmfnOl4rT9ZfRI7mLTdk1wblIUnrIq35nqwEvC"
  crossorigin="anonymous"
/>
```

The documentation for "External Stylesheets" just mentions how to import CSS from node modules.

### Additional information

```js
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 19.8.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.3.2
      eslint-config-next: 13.3.2
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by devspaceship** [comment, marked as answer]
According to the [documentation](https://nextjs.org/docs/app/building-your-application/styling/css-modules#external-stylesheets) external stylesheets should be downloaded on your server.

To fix this, you can run `npm install katex` and in the `page.tsx` where you are trying to display math, you can add a line with `import "katex/dist/katex.min.css";`

**Reply by timtucker-dte** [reply]
Not being able to refer to external stylesheets hosted in a CDN seems crazy.

While optimization of page load times is nice, it's often just as important (or more important) to optimize hosting costs by having fewer / smaller files on your own server.

**Comment by zhoushan1** [comment]
Is there a solution to this problem?

**Reply by zhoushan1** [reply]
This doesn't work in my project:

<Head>
        <link
          rel="stylesheet"
          type="text/css"
          href="https://prod-waitlist-widget.s3.us-east-2.amazonaws.com/getwaitlist.min.css"
        />
      </Head>

**Comment by judaproteam** [comment]
its not an **answer**, 
the error is: Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "https:" URIs.
how to fix it?

**Comment by stabildev** [comment]
Old discussion but this works for me:

```tsx
// layout.tsx

<html lang="en">
  <head>
    <link rel="stylesheet" href="..." />
  </head>
  <body>
    ...
```

All the metadata is still being exported using the Next APIs.

Edit: This applies global styles at the root layout

**Comment by olayway** [comment]
Why is this marked as answered? If you have a working example, can you please share?

**Comment by olayway** [comment]
That's not a very helpful information either 😔

https://nextjs.org/docs/app/building-your-application/styling/css-modules#external-stylesheets

![image](https://github.com/vercel/next.js/assets/52197250/7f978ca3-0383-4597-9424-4cdcce4e4df7)

**Reply by Konadu-Akwasi-Akuoko** [reply]
> That's not a very helpful information either 😔
> 
> https://nextjs.org/docs/app/building-your-application/styling/css-modules#external-stylesheets
> 
> ![image](https://private-user-images.githubusercontent.com/52197250/324829012-7f978ca3-0383-4597-9424-4cdcce4e4df7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0NzM4NTQsIm5iZiI6MTczNzQ3MzU1NCwicGF0aCI6Ii81MjE5NzI1MC8zMjQ4MjkwMTItN2Y5NzhjYTMtMDM4My00NTk3LTk0MjQtNGNkY2NlNGU0ZGY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIxVDE1MzIzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzMDI2MTY5YTQyMTI2NjhiNGVhMzZjMmQzYmU5Yjc2ZTBkNTkwMjVkYTJjYTBlYjcyM2I4MmM4YmM2ZTY0ZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m9lYJhgHQ7kxOB_AnbQl1tDX6w9_kLc965AtTzdSbXg)

Now we can't find this in the docs:
<img width="1440" alt="Screenshot 2025-01-21 at 3 33 50 PM" src="https://github.com/user-attachments/assets/af78dd95-86e8-4999-a61a-7007de92d21f" />

I am using the normal html `<link rel='...' href='...' </link>` style, and it works perfect for next15

**Reply by rinsuki** [reply]
its available again: https://nextjs.org/docs/app/getting-started/css#external-stylesheets

actually `<link rel="stylesheet" href="https://..." />` seems to load external stylesheets, but **after** your app's CSS, which means if your app's styles and external styles is conflicted, external styles will win, it may causes problem.

**Reply by TechQuery** [reply]
> I am using the normal html `<link rel='...' href='...' </link>` style, and it works perfect for next15

It's a benefit of React 19: https://react.dev/reference/react-dom/components/link

**Reply by megastels** [reply]
> its available again: https://nextjs.org/docs/app/getting-started/css#external-stylesheets
> 
> actually `<link rel="stylesheet" href="https://..." />` seems to load external stylesheets, but **after** your app's CSS, which means if your app's styles and external styles is conflicted, external styles will win, it may causes problem.

as a workaround
```
import ReactDOM from 'react-dom';

function layout(){
  ReactDOM.preinit('data:text/css;base64,', { as: 'style', precedence: 'high' });
  
  return (
    <html>
      <head>
        <link precedence="high" href="..." rel="..." />
      </head>
      <body>...</body>
    </html>
  );
}
<html>
```

preinit will append link with empty css, but reserve precedence `high`.
after that, all styles with precedence `high` will append before app css
